corehq/apps/analytics/utils.py

```diff
@@ -83,3 +121,173 @@ def get_instance_string():
+                return False
+            req.raise_for_status()
+        except (ConnectionError, requests.exceptions.HTTPError):
+            metrics_gauge(
+                'commcare.hubspot_data.retry.delete_hubspot_contact',
+                1
+            )
+            if retry_num <= MAX_API_RETRIES:
+                return _delete_hubspot_contact(vid, retry_num + 1)
+        else:
+            return True
+    return False
+
+
+def _get_contact_ids_for_emails(list_of_emails, retry_num=0):
+    """
+    Gets a list of Contact IDs on Hubspot from a list of emails.
+    If an email in the list doesn't exist on Hubspot, it's simply ignored.
+    :param list_of_emails:
+    :param retry_num: the number of the current retry attempt
+    :return: list of contact ids
+    """
+    if retry_num > 0:
+        time.sleep(10)  # wait 10 seconds if this is another retry attempt
+
+    api_key = settings.ANALYTICS_IDS.get('HUBSPOT_API_KEY', None)
+    if api_key:
+        try:
+            req = requests.get(
+                "https://api.hubapi.com/contacts/v1/contact/emails/batch/",
+                params={
+                    'hapikey': api_key,
+                    'email': list_of_emails,
+                },
+            )
+            if req.status_code == 404:
+                return []
+            req.raise_for_status()
+        except (ConnectionError, requests.exceptions.HTTPError):
+            metrics_gauge(
+                'commcare.hubspot_data.retry.get_contact_ids_for_emails',
+                1
+            )
+            if retry_num <= MAX_API_RETRIES:
+                return _get_contact_ids_for_emails(list_of_emails, retry_num + 1)
+        else:
+            return req.json().keys()
+    return []
+
+
+def _get_contact_ids_for_email_domain(email_domain, retry_num=0):
+    """
+    Searches Hubspot for an email domain and returns the list of matching
+    contact IDs for that email domain.
+    :param email_domain:
+    :param retry_num: the number of the current retry attempt
+    :return: list of matching contact IDs
+    """
+    if retry_num > 0:
+        time.sleep(10)  # wait 10 seconds if this is another retry attempt
+
+    api_key = settings.ANALYTICS_IDS.get('HUBSPOT_API_KEY', None)
+    if api_key:
+        try:
+            req = requests.get(
+                "https://api.hubapi.com/contacts/v1/search/query",
+                params={
+                    'hapikey': api_key,
+                    'q': f'@{email_domain}',
+                },
+            )
+            if req.status_code == 404:
+                return []
+            req.raise_for_status()
+        except (ConnectionError, requests.exceptions.HTTPError):
+            metrics_gauge(
+                'commcare.hubspot_data.retry.get_contact_ids_for_email_domain',
+                1
+            )
+            if retry_num <= MAX_API_RETRIES:
+                return _get_contact_ids_for_email_domain(email_domain, retry_num + 1)
```

Comment:
Should this `raise` on too many retries?
```suggestion
                return _get_contact_ids_for_email_domain(email_domain, retry_num + 1)
            raise
```

Reply:
see my above comment